ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -821,20 +820,40 @@ let migrate_send'  ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~options =
   eject_cds __context cd_vbds;
 
   try
+    (* Sort VDIs by size in principle and then age secondly. This gives better
+       chances that similar but smaller VDIs would arrive comparatively
+       earlier, which can serve as base for incremental copying the larger
+       ones. *)
+    let compare_fun v1 v2 =
+      let r = Int64.compare v1.size v2.size in
+      if r = 0 then
+        let t1 = Date.to_float (Db.VDI.get_snapshot_time ~__context ~self:v1.vdi) in
+        let t2 = Date.to_float (Db.VDI.get_snapshot_time ~__context ~self:v2.vdi) in
+        compare t1 t2
+      else r in
+    let all_vdis = List.sort compare_fun (suspends_vdis @ snapshots_vdis @ vms_vdis) in
+    let total_size = List.fold_left (fun acc vconf -> Int64.add acc vconf.size) 0L all_vdis in
     let so_far = ref 0L in
 
-    let with_mirrored_vdis = with_many (vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far total_size copy) in
     let new_vm =
-      with_mirrored_vdis suspends_vdis @@ fun suspends_map ->
-      with_mirrored_vdis snapshots_vdis @@ fun snapshots_map ->
-      with_mirrored_vdis vms_vdis @@ fun vdi_map ->
+      with_many (vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far total_size copy) all_vdis @@ fun all_map ->
+
+      let was_from vmap = List.exists (fun vconf -> vconf.vdi = vmap.mr_local_vdi_reference) in
+
+      let suspends_map, snapshots_map, vdi_map = List.fold_left (fun (suspends, snapshots, vdis) vmap ->
+          if was_from vmap suspends_vdis then  vmap :: suspends, snapshots, vdis
+          else if was_from vmap snapshots_vdis then suspends, vmap :: snapshots, vdis
+          else suspends, snapshots, vmap :: vdis
+        ) ([],[],[]) all_map in
+
+      let all_map = suspends_map @ snapshots_map @ vdi_map in
```

Comment:
I think `all_map = List.concat [ suspends_map; snapshots_map; vdi_map ]` would be clearer. I probably would also call it simply `maps`.